// worker-lib/dist/esm/index.js
import { createWorkerPool, initUniversalWorker } from "./pool";
const w = Worker;
export { w as Worker };
const wrapWorker = (worker) => ({
    postMessage: (message, transfer) => worker.postMessage(message, transfer),
    addEventListener: (type, listener) => {
        const handler = (e) => {
            if (e instanceof MessageEvent)
                listener(e.data);
        };
        listener._handler = handler;
        worker.addEventListener(type, handler);
    },
    removeEventListener: (type, listener) => {
        const handler = listener._handler;
        worker.removeEventListener(type, handler);
    },
    terminate: () => worker.terminate(),
});
const init = (worker) => {
    return new Promise((resolve) => {
        const handler = (e) => {
            if (e.data === "ready") {
                worker.removeEventListener("message", handler);
                resolve(worker);
            }
        };
        worker.addEventListener("message", handler);
    });
};
export const createWorker = (builder, limit = 4) => {
    return createWorkerPool(async () => {
        const result = builder();
        const worker = result instanceof Worker ? result : new Worker(result);
        await init(worker);
        return wrapWorker(worker);
    }, limit);
};
export const initWorker = (WorkerProc) => {
    const worker = self;
    const universalWorker = {
        postMessage: (message, transfer) => worker.postMessage(message, transfer),
        addEventListener: (type, listener) => {
            const handler = (e) => {
                if (e instanceof MessageEvent)
                    listener(e.data);
            };
            listener._handler = handler;
            worker.addEventListener(type, handler);
        },
        removeEventListener: (type, listener) => {
            const handler = listener._handler;
            worker.removeEventListener(type, handler);
        },
        terminate: () => {
            /* cannot terminate self */
        },
    };
    return initUniversalWorker(universalWorker, WorkerProc);
};
