// worker-lib/dist/esm/node.js
import { parentPort, Worker as WorkerNode } from "node:worker_threads";
import { createWorkerPool, initUniversalWorker } from "./pool";
export { Worker } from "node:worker_threads";
const wrapWorker = (worker) => ({
    postMessage: (message, transfer) => worker.postMessage(message, transfer),
    addEventListener: (type, listener) => {
        worker.on(type, listener);
    },
    removeEventListener: (type, listener) => {
        worker.removeListener(type, listener);
    },
    terminate: () => worker.terminate(),
});
const init = (worker) => {
    return new Promise((resolve) => {
        worker.once("message", () => {
            resolve(worker);
        });
    });
};
export const createWorker = (builder, limit = 4) => {
    return createWorkerPool(async () => {
        const result = builder();
        const worker = result instanceof WorkerNode ? result : new WorkerNode(result);
        await init(worker);
        return wrapWorker(worker);
    }, limit);
};
export const initWorker = (WorkerProc) => {
    const worker = parentPort;
    if (!worker) {
        throw new Error("This is not a worker thread");
    }
    const universalWorker = {
        postMessage: (message, transfer) => worker.postMessage(message, transfer),
        addEventListener: (type, listener) => {
            worker.on(type, listener);
        },
        removeEventListener: (type, listener) => {
            worker.removeListener(type, listener);
        },
        terminate: () => {
            /* cannot terminate self */
        },
    };
    return initUniversalWorker(universalWorker, WorkerProc);
};
