// worker-lib/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initWorker = exports.createWorker = exports.Worker = void 0;
const pool_1 = require("./pool");
const w = Worker;
exports.Worker = w;
const wrapWorker = (worker) => ({
    postMessage: (message, transfer) => worker.postMessage(message, transfer),
    addEventListener: (type, listener) => {
        const handler = (e) => {
            if (e instanceof MessageEvent)
                listener(e.data);
        };
        listener._handler = handler;
        worker.addEventListener(type, handler);
    },
    removeEventListener: (type, listener) => {
        const handler = listener._handler;
        worker.removeEventListener(type, handler);
    },
    terminate: () => worker.terminate(),
});
const init = (worker) => {
    return new Promise((resolve) => {
        const handler = (e) => {
            if (e.data === "ready") {
                worker.removeEventListener("message", handler);
                resolve(worker);
            }
        };
        worker.addEventListener("message", handler);
    });
};
const createWorker = (builder, limit = 4) => {
    return (0, pool_1.createWorkerPool)(async () => {
        const result = builder();
        const worker = result instanceof Worker ? result : new Worker(result);
        await init(worker);
        return wrapWorker(worker);
    }, limit);
};
exports.createWorker = createWorker;
const initWorker = (WorkerProc) => {
    const worker = self;
    const universalWorker = {
        postMessage: (message, transfer) => worker.postMessage(message, transfer),
        addEventListener: (type, listener) => {
            const handler = (e) => {
                if (e instanceof MessageEvent)
                    listener(e.data);
            };
            listener._handler = handler;
            worker.addEventListener(type, handler);
        },
        removeEventListener: (type, listener) => {
            const handler = listener._handler;
            worker.removeEventListener(type, handler);
        },
        terminate: () => {
            /* cannot terminate self */
        },
    };
    return (0, pool_1.initUniversalWorker)(universalWorker, WorkerProc);
};
exports.initWorker = initWorker;
