// worker-lib/dist/cjs/node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initWorker = exports.createWorker = exports.Worker = void 0;
const node_worker_threads_1 = require("node:worker_threads");
const pool_1 = require("./pool");
var node_worker_threads_2 = require("node:worker_threads");
Object.defineProperty(exports, "Worker", { enumerable: true, get: function () { return node_worker_threads_2.Worker; } });
const wrapWorker = (worker) => ({
    postMessage: (message, transfer) => worker.postMessage(message, transfer),
    addEventListener: (type, listener) => {
        worker.on(type, listener);
    },
    removeEventListener: (type, listener) => {
        worker.removeListener(type, listener);
    },
    terminate: () => worker.terminate(),
});
const init = (worker) => {
    return new Promise((resolve) => {
        worker.once("message", () => {
            resolve(worker);
        });
    });
};
const createWorker = (builder, limit = 4) => {
    return (0, pool_1.createWorkerPool)(async () => {
        const result = builder();
        const worker = result instanceof node_worker_threads_1.Worker ? result : new node_worker_threads_1.Worker(result);
        await init(worker);
        return wrapWorker(worker);
    }, limit);
};
exports.createWorker = createWorker;
const initWorker = (WorkerProc) => {
    const worker = node_worker_threads_1.parentPort;
    if (!worker) {
        throw new Error("This is not a worker thread");
    }
    const universalWorker = {
        postMessage: (message, transfer) => worker.postMessage(message, transfer),
        addEventListener: (type, listener) => {
            worker.on(type, listener);
        },
        removeEventListener: (type, listener) => {
            worker.removeListener(type, listener);
        },
        terminate: () => {
            /* cannot terminate self */
        },
    };
    return (0, pool_1.initUniversalWorker)(universalWorker, WorkerProc);
};
exports.initWorker = initWorker;
