// wasm-image-optimization/dist/core.js
export var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["None"] = 0] = "None";
    LogLevel[LogLevel["Error"] = 1] = "Error";
    LogLevel[LogLevel["Warning"] = 2] = "Warning";
    LogLevel[LogLevel["Info"] = 3] = "Info";
    LogLevel[LogLevel["Debug"] = 4] = "Debug";
})(LogLevel || (LogLevel = {}));
export var RenderFormat;
(function (RenderFormat) {
    RenderFormat[RenderFormat["None"] = -1] = "None";
    RenderFormat[RenderFormat["SVG"] = 0] = "SVG";
    RenderFormat[RenderFormat["PNG"] = 1] = "PNG";
    RenderFormat[RenderFormat["WebP"] = 2] = "WebP";
    RenderFormat[RenderFormat["PDF"] = 3] = "PDF";
    RenderFormat[RenderFormat["JPEG"] = 4] = "JPEG";
    RenderFormat[RenderFormat["AVIF"] = 5] = "AVIF";
    RenderFormat[RenderFormat["RAW"] = 6] = "RAW";
    RenderFormat[RenderFormat["ThumbHash"] = 7] = "ThumbHash";
})(RenderFormat || (RenderFormat = {}));
export var FitMode;
(function (FitMode) {
    FitMode[FitMode["Contain"] = 0] = "Contain";
    FitMode[FitMode["Cover"] = 1] = "Cover";
    FitMode[FitMode["Fill"] = 2] = "Fill";
})(FitMode || (FitMode = {}));
export class ImageConverterBase {
    factory;
    modPromise;
    _onLog;
    _logLevel = LogLevel.None;
    constructor(factory) {
        this.factory = factory;
    }
    get onLog() {
        return this._onLog;
    }
    set onLog(val) {
        this._onLog = val;
        this.getModule().then((mod) => {
            mod.onLog = val;
        });
    }
    get logLevel() {
        return this._logLevel;
    }
    set logLevel(val) {
        this._logLevel = val;
        this.getModule().then((mod) => {
            mod.logLevel = val;
        });
    }
    async getModule() {
        if (!this.modPromise) {
            this.modPromise = (async () => {
                let currentLogLevel = this._logLevel;
                let currentUserOnLog = this._onLog;
                const initialModule = {
                    onLog: (level, message) => {
                        if (this._logLevel !== LogLevel.None &&
                            level <= this._logLevel &&
                            this._onLog) {
                            this._onLog(level, message);
                        }
                    },
                    print: (text) => {
                        if (this._logLevel !== LogLevel.None &&
                            LogLevel.Info <= this._logLevel &&
                            this._onLog) {
                            this._onLog(LogLevel.Info, text);
                        }
                    },
                    printErr: (text) => {
                        if (this._logLevel !== LogLevel.None &&
                            LogLevel.Error <= this._logLevel &&
                            this._onLog) {
                            this._onLog(LogLevel.Error, text);
                        }
                    },
                };
                const mod = (await this.factory(initialModule));
                mod.set_log_level(this._logLevel);
                // In modularized mode, the factory returns the module object.
                // EM_JS might be looking for functions on this object.
                mod.onLog = initialModule.onLog;
                mod.logLevel = this._logLevel;
                const originalSetLogLevel = mod.set_log_level;
                mod.set_log_level = (level) => {
                    this._logLevel = level;
                    originalSetLogLevel(level);
                };
                let internalOnLog = mod.onLog;
                Object.defineProperty(mod, "onLog", {
                    get: () => internalOnLog,
                    set: (val) => {
                        internalOnLog = val;
                        this._onLog = val;
                    },
                });
                let internalLogLevel = mod.logLevel;
                Object.defineProperty(mod, "logLevel", {
                    get: () => internalLogLevel,
                    set: (val) => {
                        internalLogLevel = val;
                        this._logLevel = val;
                    },
                });
                return mod;
            })();
        }
        return this.modPromise;
    }
    async setLogLevel(level) {
        this.logLevel = level;
    }
    async optimizeImage(params) {
        const mod = await this.getModule();
        const inst = mod.create_instance();
        const { image: imageInput, width, height, fit = "contain", format, quality, speed = 6, animation = false, crop, } = params;
        const image = imageInput instanceof Uint8Array
            ? imageInput
            : typeof imageInput === "string"
                ? new TextEncoder().encode(imageInput)
                : new Uint8Array(imageInput);
        try {
            if (!mod.load_image(inst, image)) {
                throw new Error("Failed to load image");
            }
            const originalWidth = mod.get_original_width(inst);
            const originalHeight = mod.get_original_height(inst);
            const originalAnimation = mod.is_original_animation(inst);
            const originalFormat = mod.get_original_format(inst);
            let targetFormat = format || "webp";
            if (targetFormat !== "none" && animation && originalAnimation) {
                targetFormat = "webp";
            }
            if (crop) {
                mod.crop(inst, crop.x, crop.y, crop.width, crop.height);
            }
            if (targetFormat !== "none" && (width || height)) {
                const fitMap = {
                    contain: FitMode.Contain,
                    cover: FitMode.Cover,
                    fill: FitMode.Fill,
                };
                const fMode = fitMap[fit] ?? FitMode.Contain;
                mod.resize(inst, width ?? 0, height ?? 0, fMode);
            }
            const formatMap = {
                none: RenderFormat.None,
                png: RenderFormat.PNG,
                webp: RenderFormat.WebP,
                jpeg: RenderFormat.JPEG,
                avif: RenderFormat.AVIF,
                raw: RenderFormat.RAW,
                thumbhash: RenderFormat.ThumbHash,
            };
            const f = formatMap[targetFormat] ?? RenderFormat.WebP;
            const result = mod.encode(inst, f, quality ?? 100, speed, animation);
            if (!result) {
                throw new Error("Failed to encode image");
            }
            return {
                data: new Uint8Array(result),
                originalWidth,
                originalHeight,
                originalAnimation,
                originalFormat,
                width: mod.get_width(inst),
                height: mod.get_height(inst),
                animation: mod.is_animation(inst),
                format: mod.get_format(inst),
            };
        }
        finally {
            mod.destroy_instance(inst);
        }
    }
}
